### ISSUE TITLE
Remove dependency on local build folder

### ISSUE THREAD

**Issue by pabelanger** [ansible/ansible-builder#184]
I know I've raised this point before, but I feel now that we have multi-stage builds working properly, we can finally work the idea of removing the build stage mounts for introspect.

Here is a POC (non-working) Dockerfile that should explain what I am thinking.

```
ARG ANSIBLE_RUNNER_IMAGE=quay.io/ansible/ansible-runner:devel
ARG PYTHON_BUILDER_IMAGE=quay.io/ansible/python-builder:latest

FROM $ANSIBLE_RUNNER_IMAGE as galaxy

ARG ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=
ADD _build /build

WORKDIR /build
RUN ansible-galaxy role install -r requirements.yml --roles-path /usr/share/ansible/roles
RUN ansible-galaxy collection install $ANSIBLE_GALAXY_CLI_COLLECTION_OPTS -r requirements.yml --collections-path /usr/share/ansible/collections

from galaxy as requirements
RUN pip install ansible-builder
WORKDIR /usr/local/lib/python3.8/site-packages/ansible_builder
RUN python3 introspect.py

FROM $PYTHON_BUILDER_IMAGE as builder
COPY --from=requirements /tmp/requirements_combined.txt /tmp/src/requirements.txt
COPY --from=requirements /tmp/bindep_combined.txt /tmp/src/bindep.txt
RUN assemble

FROM $ANSIBLE_RUNNER_IMAGE

COPY --from=galaxy /usr/share/ansible /usr/share/ansible

COPY --from=builder /output/ /output/
RUN /output/install-from-bindep && rm -rf /output/wheels

```

The idea here is, ansible-builder does some special magic of creating 2 dockerfile to produce an EE, expect we overwrite the first. We basically have 2 stages of ansible-builder, where if you run podman directly on the dockerfile, it isn't the same process.

The purpose of phase 1, is to install collections, introspect and create build time files. Then phase 2 consumes those files.  But, the final dockerfile, does not expose how phase 1 works.  This is magic inside ansible-builder today.

I given we now have multi-stage builds working, I believe we can create a new layer, where we install ansible-builder introspect files, and produce the combined requirements.txt / bindep.txt files, then pass them into the next layer using --from flag in COPY command.

This would mean, more logic inside of Dockerfile (this could be minimized like our assemble script) and helps to remove some magic hidden inside ansible-builder.

This also solves the issue, where an ansible collection, that updates requirements.txt or bindep.txt files, will always have to re-run ansible-builder to sync the _build stage directory. Cutting down on the frequency a user has to regenerate ansible-builder, for requirement changes.

**Comment by pabelanger** [comment]
Basically, we first have to land https://github.com/ansible/network-ee/pull/33, moving collection requirements into the network-ee, to merge https://github.com/ansible-collections/ansible.utils/pull/38, then we will revert and regenerate network-ee with ansible-builder to ensure our build context is up to date.

**Comment by AlanCoding** [comment]
I don't understand the need to have `from galaxy as requirements`. Just keep working in that image as `galaxy`. That already doesn't become part of the final build. I just don't see the point of introducing a new image.

```
RUN pip install ansible-builder
WORKDIR /usr/local/lib/python3.8/site-packages/ansible_builder
RUN python3 introspect.py
```

can be written more simply as:

```
RUN pip install ansible-builder
RUN ansible-builder introspect
```

There's more to discuss, I just want to help move things forward to touching up the mockup a bit.

**Comment by pabelanger** [comment]
> I don't understand the need to have `from galaxy as requirements`. Just keep working in that image as `galaxy`. That already doesn't become part of the final build. I just don't see the point of introducing a new image.
> 
> ```
> RUN pip install ansible-builder
> WORKDIR /usr/local/lib/python3.8/site-packages/ansible_builder
> RUN python3 introspect.py
> ```
> 
> can be written more simply as:
> 
> ```
> RUN pip install ansible-builder
> RUN ansible-builder introspect
> ```
> 
> There's more to discuss, I just want to help move things forward to touching up the mockup a bit.

100%, agree. THis was just v1 of the POC, galaxy layer is disposible so we can also include introspect there

**Comment by AlanCoding** [comment]
I don't follow the title of this issue. The Containerfile still has `ADD _build /build`, which may contain an `ansible.cfg` with Galaxy server config and/or collection tarballs. That won't be removed.

It's not technically problematic to install `ansible-builder` into the container image, but this means you are carrying 2 copies of `ansible-builder`, one on the host machine and one inside the container. That makes life complicated, and I'd prefer a canonical process that for-sure uses 1 copy of this software that's just in 1 layer of the stack.

`ansible-builder compile` used to be a command which produced the build context & Containerfile but didn't call `podman build`. This became non-viable with the introduction of 2-phase builds, and your proposal reverses that. If the primary use case is to run `introspect` inside the container, then this should be considered a tool to be _used by_ podman, not a tool _that runs_ podman.

I went ahead and made https://github.com/ansible/ansible-builder/pull/186, with the idea of making it even less picky to install.

**Comment by pabelanger** [comment]
You are right, for the reasons above, removing build directory isn't correct. It is more the 2 phase build process and using the local filesystem to store the requirements / bindep combined files.  That really causes me some pain, while trying to rebuild an EE, if a collection changes those files pre-relase.

**Comment by AlanCoding** [comment]
I want us to keep in mind that the introspect logic is important as a source of truth for what winds up in an execution environment, and this is intended to fit into a Galaxy integration at some point. This is the one consumer we have:

https://github.com/ansible/galaxy-importer/blob/ed96acbf660102eb5ebef26f857abece584ae525/galaxy_importer/utils/execution_environment.py#L18

If we move the introspect script, we should be decisive about it. Lots of our tests in `ansible-builder` depend on its behavior and that can't be refactored out. Hypothetically, say it moved to `ansible-runner`, we might still be able to make changes and run tests using the Depends-On markers.

**Comment by AlanCoding** [comment]
I'm trying to fold in all the high-level topics we talked about, and get the most basic visibility of what we're thinking about.

```
ARG ANSIBLE_RUNNER_IMAGE=quay.io/ansible/ansible-runner:devel
ARG PYTHON_BUILDER_IMAGE=quay.io/ansible/python-builder:latest

FROM $ANSIBLE_RUNNER_IMAGE as galaxy

ARG ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=
ADD _build /build

WORKDIR /build
RUN ansible-galaxy role install -r requirements.yml --roles-path /usr/share/ansible/roles
RUN ansible-galaxy collection install $ANSIBLE_GALAXY_CLI_COLLECTION_OPTS -r requirements.yml --collections-path /usr/share/ansible/collections

RUN pip install ansible-builder
RUN ansible-builder introspect --write-bindep=/tmp/src/bindep.txt --write-python=/tmp/src/requirements.txt

FROM $PYTHON_BUILDER_IMAGE as builder
COPY --from=galaxy /tmp/src/requirements.txt /tmp/src/requirements.txt
COPY --from=galaxy /tmp/src/bindep.txt /tmp/src/bindep.txt
RUN assemble

FROM $ANSIBLE_RUNNER_IMAGE

COPY --from=galaxy /usr/share/ansible /usr/share/ansible

COPY --from=builder /output/ /output/
RUN /output/install-from-bindep && rm -rf /output/wheels
```

This might be getting closer, but a lot doesn't feel right. The entire existence of the 2nd stage is awkward. The only reason that image is more than a collection of bash scripts is for `bindep`, but if we're installing ansible-builder ad-hoc, why are we not doing the same for bindep? The COPYs from galaxy are awkward too.

Even though the folder is large, I think we should prefer to COPY `/usr/share/ansible` from `galaxy` to `builder`, then in `builder` install `ansible-builder`, make `bindep` a dependency of `ansible-builder`, and then incorporate the scripts as sub-commands to `ansible-builder`. This, to me, looks like something that can make the Dockerfile less hard to follow.

**Comment by shanemcd** [comment]
This is done via #217
